7203 - Alchemy rpc provider

MRLab12 opened the pull request:
## Description

<s>_I have opened this Draft PR to make sure I did the correct changes and get some feedback since this is my first contribution here._</s>

This PR adds an Alchemy provider in `web3-rpc-providers`. I matched the design of the Quicknode provider. 

For testing, I wasn't sure if a test for this provider should actually try to make a request. Should we consider mocking? I want to learn more about how testing is done in web3.js.


Fixes #7203 

## Type of change

<!-- Please delete options that are not relevant. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [x] I ran `npm run lint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:unit` with success.
- [ ] I ran `npm run test:coverage` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.
- [x] I have linked Issue(s) with this PR in "Linked Issues" menu.

MRLab12 committed:
Add alchemy rpc provider

MRLab12 committed:
Add Polygon AMOY network (replaces MUMBAI)

MRLab12 committed:
Keep this for now, check what is using this property

MRLab12 committed:
Testing

jdevcs commented on packages/web3-rpc-providers/src/web3_provider_alchemy.ts:
```diff
@@ -0,0 +1,124 @@
+		super(network, transport, token, host, providerConfigOptions);
+	}
+
+	public async request<
+		Method extends Web3APIMethod<API>,
+		ResultType = Web3APIReturnType<API, Method>,
+	>(
+		payload: Web3APIPayload<EthExecutionAPI, Method>,
+		requestOptions?: RequestInit,
+	): Promise<JsonRpcResponseWithResult<ResultType>> {
+		try {
+			return await super.request(payload, requestOptions);
+		} catch (error) {
+			if (error instanceof ResponseError && error.statusCode === 429) {
+				throw new Error('Alchemy rate limit exceeded');
+			}
+			throw error;
+		}
+	}
```
```suggestion

```
As currently there is no custom logic for this provider. 429 status code is specific for QN provider

jdevcs commented on packages/web3-rpc-providers/src/web3_provider_alchemy.ts:
```diff
@@ -0,0 +1,124 @@
+		let host = '';
+		let token = '';
+
+		switch (network) {
```
instead of switch use map as https://github.com/web3/web3.js/pull/7260/files#diff-b3d860d130c02e49ecd127d925384fe54f14e0b93c94aec96cf2965781c9efa8R37

MRLab12 committed:
Remove request method

Co-authored-by: Junaid <86780488+jdevcs@users.noreply.github.com>

MRLab12 committed:
Updating imports and unused API extension / Use map instead of switch

MRLab12 committed:
Fix import in rpc tests

avkos commented on packages/web3-rpc-providers/src/web3_provider_alchemy.ts:
```diff
@@ -0,0 +1,61 @@
+		[Network.POLYGON_AMOY]: 'polygon-amoy.g.alchemy.com',
+		[Network.OPTIMISM_MAINNET]: 'opt-mainnet.g.alchemy.com',
+		[Network.OPTIMISM_SEPOLIA]: 'opt-sepolia.g.alchemy.com',
+	};
```
It would be great to add all Alchemy platforms if they are compatible with web3js.

<img width="885" alt="image" src="https://github.com/user-attachments/assets/29623525-f25f-4ca9-a21e-36e0672317c0">

MRLab12 replied:
Is this screenshot from web.js? Wondering if this is an accurate list of the chains to support with Alchemy.

MRLab12 replied:
@avkos Made some changes, not sure if all the chains I added are to be supported but lmk.

avkos commented on packages/web3/test/integration/web3RPCProviders.test.ts:
```diff
@@ -42,6 +48,18 @@ describe('Web3 RPC Provider Integration tests', () => {
+					web3.provider?.disconnect();
+				}
+			});
+			it(`AlchemyProvider should work with ${transport} transport and ${network} network`, async () => {
```
After adding all platforms it will be a different array of networks. so it will be better to have separate const `networksAlchemy` and separate test `networksAlchemy.forEach`

MRLab12 committed:
Include more Alchemy networks

MRLab12 committed:
Make a separate test loop for Alchemy networks and use a specific network list

MRLab12 committed:
Merge branch '4.x' into alchemy-rpc-provider

# Conflicts:
#	packages/web3-rpc-providers/src/types.ts
#	packages/web3/test/integration/web3RPCProviders.test.ts

MRLab12 committed:
ETH Goerli has been removed

Muhammad-Altabba committed:
Merge branch '4.x' into alchemy-rpc-provider

Muhammad-Altabba reviewed (approved):
Many thanks @MRLab12 for your contribution,
It seems great now. I think, once other team members approve, we can merge this MR.

MRLab12 commented:
I'm trying to run integration tests locally but having issues. Looking at the test errors here, most likely a missing Alchemy key is the issue. Can someone point me to how this should be set up?

jdevcs commented:
> I'm trying to run integration tests locally but having issues. Looking at the test errors here, most likely a missing Alchemy key is the issue. Can someone point me to how this should be set up?

For your local testing I'll suggest to use your personal key, and for web3.js one of our team member will add web3.js specific alchemy key.

MRLab12 commented:
> > I'm trying to run integration tests locally but having issues. Looking at the test errors here, most likely a missing Alchemy key is the issue. Can someone point me to how this should be set up?
> 
> For your local testing I'll suggest to use your personal key, and for web3.js one of our team member will add web3.js specific alchemy key.

Hey! Yup, I used my personal key to manually test the functionality using a script. For the tests I added, I thought these tests were going to be ignored for now. Since it looks like they are being used, we will most likely need to set up the key for the tests to work.

I'm going to try to get the tests I added to run locally, but when I last tried using `WEB3_SYSTEM_TEST_BACKEND=geth WEB3_SYSTEM_TEST_PROVIDER=ws://127.0.0.1:8545 yarn test:integration`, tests seem to work but it also generates some errors.  Maybe someone can help me out with running newly added tests locally without running all the test suites.

Muhammad-Altabba committed:
Merge branch '4.x' into alchemy-rpc-provider

Muhammad-Altabba committed:
set a temp default Alchemy api token

Muhammad-Altabba committed:
revert: set a temp default Alchemy api token

Muhammad-Altabba committed:
Merge branch '4.x' into alchemy-rpc-provider

MRLab12 committed:
Merge branch '4.x' into alchemy-rpc-provider

jdevcs commented:
Thank you for your PR. We appreciate your contributions to the Web3.js library.
As you may be aware, we are in the process of sunsetting Web3.js, and as such, we will not be merging any non-critical pull requests. 
Our focus during this transition will be on addressing critical issues to ensure a smooth migration for our users.
For further details regarding the sunsetting of Web3.js, please refer to the official announcement here: [Web3.js Sunset Announcement](https://blog.chainsafe.io/web3-js-sunset/)
